### PR TITLE
Curvature resistance in train dynamics

### PR DESCRIPTION
**Pull request by JuxhinoKavaja**
With this pull request we intoduce curvature resistance to the train dynamics.

### CODE REVIEW

File: track.py [hunk 1]
```diff
@@ -179,8 +186,15 @@ def __init__(self, config, pathJSON='tracks'):
         self.title = data['metadata']['id']
 
         self.importSpeedLimitTuples(data['speed limits']['values'], data['speed limits']['units']['velocity'])
+
         self.importGradientTuples(data['gradients']['values'], data['gradients']['units']['slope'])
 
+        self.importCurvatureTuples(data['curvatures']['values'] if 'curvatures' in data else [(0.0, "infinity", "infinity")],
```
Comment (dkouzoup) [inline track.py]: Nice. In principle, gradients are also optional, could you apply the same logic there? I forgot to do so.

File: track.py [hunk 2]
```diff
@@ -266,6 +293,96 @@ def importSpeedLimitTuples(self, tuples, unit='km/h'):
         checkDataFrame(self.speedLimits, self.length)
 
 
+    def importCurvatureTuples(self, tuples, unitRadiusStart='m', unitRadiusEnd='m', clothoidSamplingInterval=None):
+
+        if not self.lengthOk():
+
+            raise ValueError("Cannot import curvature without a valid track length!")
+        
+        if not unitRadiusStart == unitRadiusEnd:
```
Comment (dkouzoup) [inline track.py]: I understand that it barely make sense to define different units for radius at start and at end, but since we provide two fields for these two values, it should be supported (i.e., radius at start in m and radius at end in km). 

The convertUnit function will take care of converting both to m. Also the check in line 308 not necessary since it happens already inside the convertUnit function.

File: track.py [hunk 3]
```diff
@@ -266,6 +293,96 @@ def importSpeedLimitTuples(self, tuples, unit='km/h'):
...
+
+                sectionEnd = tuples[trackIndex+1][0] if trackIndex < len(tuples)-1 else self.length
+
+                if ds == None or int((sectionEnd-sectionStart)/ds) == 0:
+
+                    result.append((sectionStart, (curvatureStart + curvatureEnd)/2))
+
+                else:
+
+                    nIntervals = int((sectionEnd-sectionStart)/ds)
```
Comment (dkouzoup) [inline track.py:377]: Do we have a unit test that this is doing the right thing?

File: track.py [hunk 4]
```diff
@@ -266,6 +293,96 @@ def importSpeedLimitTuples(self, tuples, unit='km/h'):
...
+
+        epsilon = sys.float_info.epsilon
+
+        for trackIndex, trackSection in enumerate(tuples):
+
+            sectionStart = trackSection[0]
+            curvatureStart = 1/trackSection[1]
+            curvatureEnd = 1/trackSection[2]
+
+            if (curvatureStart - curvatureEnd) <= epsilon:
```
Comment (dkouzoup) [inline track.py]: Does this work correctly with negative radius? Don't we need abs?

File: unitTests/curvatureResistance/curvatureResistance.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+    DEFAULT_SPEED_AT_END = 1 # [m/s]
+    
+    THRESHOLD_VELOCITY_ERROR_MINIMUM_TIME_PROBLEM = 1e-3 # adimensional quantity
+
+    THRESHOLD_ENERGY_ERROR_MINIMUM_ENERGY_PROBLEM = 5e-2 # adimensional quantity
+
+    @classmethod
+    def setUpClass(cls):
+        '''
+        Based on the dataset contained in TRACK, two temporary track datasets are created.
```
Comment (dkouzoup) [inline unitTests/curvatureResistance/curvatureResistance.py]: Isn't it simpler to create two track objects from the existing json file instead of creating temporary json files?

File: unitTests/curvatureResistance/curvatureResistance.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+    @classmethod
+    def tearDownClass(cls):
+
+        cls.fileNoCurvature.close(), cls.fileConstantCurvature.close()
+
+        os.unlink(cls.fileNoCurvature.name), os.unlink(cls.fileConstantCurvature.name)
+
+    
+
+    def specificCurvatureResistanceForce(cls, g, rho):
```
Comment (dkouzoup) [inline unitTests/curvatureResistance/curvatureResistance.py]: since all functions that follow are not class methods, shouldn't we use 'self' instead of 'cls' as first input?

**Review by dkouzoup — COMMENTED** [review]
Thanks, I added a few comments to be discussed. Haven't checked the unit tests in detail yet. Please make sure that the changes do not break the scripts in the simulation folder.

File: track.py (hunk 3 above)
Comment (JuxhinoKavaja) [reply track.py:377]: I added the test  with the latest commit.I will close this thread but please have a look at the unit test.

File: track.py (hunk 4 above)
Comment (JuxhinoKavaja) [reply track.py]: Good catch. It was indeed necessary. Thanks!

File: track.py [hunk 7]
```diff
@@ -266,6 +293,106 @@ def importSpeedLimitTuples(self, tuples, unit='km/h'):
...
+        of the clothoid curvature in the track section starting at position p.
+        """
+
+        if any([radiusValue == 0 for radiusValue in [trackSection[radiusType] for trackSection in tuples for radiusType in range(1,3)]]):
+
+            raise ValueError("Curvature radius can not be 0!")
+
+        if any([tuples[sectionIndex][0] < 0 for sectionIndex in range(len(tuples))]):
+
+            raise ValueError("Positions can not be negative!")
```
Comment (dkouzoup) [inline track.py]: cannot :)

File: track.py [hunk 8]
```diff
@@ -364,11 +492,12 @@ def crop(dfIn):
...
         self.gradients = crop(self.gradients)
+        self.curvatures = crop(self.curvatures)
 
 
 if __name__ == '__main__':
 
     # Example on how to load and plot a track
 
-    track = Track(config={'id':'00_stationX_stationY'})
+    track = Track(config={'id':'00_stationX_stationY', 'clothoidSamplingInterval': 10})
```
Comment (dkouzoup) [inline track.py]: why do you pass the clothoid sampling interval option if the track has no curvature information?

File: unitTests/curvatureResistance/curvatureResistance.py [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+        - param finalPosition: position where the trip must end.
+        - param realLossFunction: if provided this loss function is used to compute the real losses of the train. Note that it is not used to
+        solve the minimum energy problem.
+        - return: a dataframe containing the solution of the problem.
+        '''
+
+        v0 = self.speedAtStart
+        vN = self.speedAtEnd
+
+        if (train == None):
```
Comment (dkouzoup) [inline unitTests/curvatureResistance/curvatureResistance.py]: I would suggest to pass the train "ready" to this function instead the None check.

File: unitTests/curvatureResistance/curvatureResistance.py [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+
+            train = Train(train='Intercity')
+
+            train.forceMinPn = 0
+
+        if (finalPosition == None):
```
Comment (dkouzoup) [inline unitTests/curvatureResistance/curvatureResistance.py]: This I would also do in the test script already

File: unitTests/curvatureResistance/curvatureResistance.py [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+        track.updateLimits(positionEnd=finalPosition)
+
+        solverOpts = { "maxIterations": 500,
+                    "numIntervals": 300,
+                    "integrationMethod": "RK",
+                    "integrationOptions": {"order": 4, "numSteps": 1, "numApproxSteps": 1 },
+                    "energyOptimal": energyOptimal,
+                    "minimumVelocity": min(v0, vN)}
+
+        # # # solve problem with perfect efficiency
```
Comment (dkouzoup) [inline unitTests/curvatureResistance/curvatureResistance.py]: copy pasted comment

File: unitTests/curvatureResistance/curvatureResistance.py [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+        train.powerLosses = lossFunction
+
+        ocp0 = casadiSolver(train, track, solverOpts)
+
+        df0, _ = ocp0.solve(terminalTime, terminalVelocity=vN, initialVelocity=v0)
+
+        if (realLossFunction != None):
```
Comment (dkouzoup) [inline unitTests/curvatureResistance/curvatureResistance.py]: I would not re calculate based on "real" losses, simply check that tests pass on the loss function that the OCP knows

**Review by dkouzoup — APPROVED** [review]
Minor fixes, let me know when done and I will merge it
